src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs

```diff
@@ -0,0 +1,363 @@
+                    AcsSpFilePath));
+            }
+            StoreServicePrincipal(acsServicePrincipal);
+            return acsServicePrincipal;
+        }
+
+        private void BuildServicePrincipal(string name, string url, string clientSecret)
+        {
+            var pwCreds = new PasswordCredential(
+                value: clientSecret,
+                startDate: DateTime.UtcNow,
+                endDate: DateTime.UtcNow.AddYears(2));
+
+            var app = GraphClient.Applications.Create(new ApplicationCreateParameters(
+                false,
+                name,
+                new List<string> {url},
+                url,
+                passwordCredentials: new List<PasswordCredential> {pwCreds}));
+
+            var success = RetryAction(() =>
+            {
+                var sp = new ServicePrincipalCreateParameters(
+                    app.AppId,
+                    true,
+                    passwordCredentials: new List<PasswordCredential> {pwCreds});
+                GraphClient.ServicePrincipals.Create(sp);
+            });
+
+            if (!success)
+            {
+                throw new CmdletInvocationException(
+                    "Could not create a service principal with the right permissions. Are you an Owner on this project?");
+            }
+
+            AddRoleAssignment("Contributor", app.AppId);
+        }
+
+        protected bool Exists()
+        {
+            try
+            {
+                var exists = Client.ManagedClusters.Get(ResourceGroupName, Name) != null;
+                WriteVerbose(string.Format("Cluster exists: {0}", exists));
+                return exists;
+            }
+            catch (CloudException)
+            {
+                WriteVerbose("Cluster does not exist.");
+                return false;
+            }
+        }
+
+        protected void AddRoleAssignment(string role, string appId)
+        {
+            var success = RetryAction(() =>
+                AuthClient.RoleAssignments.Create(role, appId, new RoleAssignmentCreateParameters()
+                {
+                    Properties = new RoleAssignmentProperties(role, appId)
+                }));
+
+            if (!success)
+            {
+                throw new CmdletInvocationException(
+                    "Could not create a service principal with the right permissions. Are you an Owner on this project?");
+            }
+        }
+
+        protected static bool RetryAction(Action action)
+        {
+            var success = false;
+            foreach (var i in Enumerable.Range(1, 10))
+            {
+                try
+                {
+                    action();
+                    success = true;
+                    break;
+                }
+                catch (Exception)
+                {
+                    // AAD might puke here, so we catch it and try again until success
+                    Thread.Sleep(1000 * i);
```

Comment:
Use [TestMockSupport.Delay](https://github.com/Azure/azure-powershell/blob/preview/src/Common/Commands.Common/TestMockSupport.cs#L34) instead of calling Thread.Sleep directly

Reply:
This is not a test. I repeat this is not a test. 

No, but seriously, this needs to be here because sometimes we need to wait a bit for AAD to propagate changes so that we can add the ARM RBAC role to the service principal.